subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -1110,376 +644,465 @@ static void fota_evt_handler(struct azure_fota_event *fota_evt)
-	err = azure_fota_init(fota_evt_handler);
+	err = mqtt_helper_init(&cfg);
 	if (err) {
-		LOG_ERR("Failed to initialize Azure FOTA, error: %d", err);
-		return err;
+		LOG_ERR("mqtt_helper_init failed, error: %d", err);
+		return -EFAULT;
 	}
 
-	LOG_DBG("Azure FOTA initialized");
-#endif
+	hostname_span = az_span_create(conn_params.hostname.ptr, conn_params.hostname.size);
+	device_id_span = az_span_create(conn_params.device_id.ptr, conn_params.device_id.size);
 
-	evt_handler = event_handler;
-	is_initialized = true;
-	(void)err;
+	/* Initialize Azure SDK client instance. */
+	err = az_iot_hub_client_init(
+		&iot_hub_client,
+		hostname_span,
+		device_id_span,
+		NULL);
+	if (az_result_failed(err)) {
+		LOG_ERR("Failed to initialize IoT Hub mqtt_client, result code: %d", err);
+		iot_hub_state_set(STATE_DISCONNECTED);
 
-	connection_state_set(STATE_INIT);
+		return -EFAULT;
+	}
+
+	err = az_iot_hub_client_get_user_name(&iot_hub_client,
+					      user_name_buf,
+					      sizeof(user_name_buf),
+					      &user_name_len);
+	if (az_result_failed(err)) {
+		LOG_ERR("Failed to get user name, az error: 0x%08x", err);
+		return -EFAULT;
+	}
+
+	conn_params.user_name.size = user_name_len;
+
+	LOG_DBG("User name: %.*s", conn_params.user_name.size, conn_params.user_name.ptr);
+	LOG_DBG("User name buffer size is %d, actual user name size is: %d",
+		sizeof(user_name_buf), user_name_len);
+
+	err = mqtt_helper_connect(&conn_params);
+	if (err) {
+		LOG_ERR("mqtt_helper_connect failed, error: %d", err);
+		return err;
+	}
 
 	return 0;
 }
 
-int azure_iot_hub_method_respond(struct azure_iot_hub_result *result)
+int azure_iot_hub_send(const struct azure_iot_hub_msg *const msg)
 {
-	ssize_t len;
-	static char topic[100];
+	int err;
+	ssize_t topic_len;
+	char topic[CONFIG_AZURE_IOT_HUB_TOPIC_MAX_LEN + 1];
+
+	if (msg == NULL) {
+		return -EINVAL;
+	}
+
 	struct mqtt_publish_param param = {
-		.message.payload.data = (char *)result->payload,
-		.message.payload.len = result->payload_len,
-		.message.topic.topic.utf8 = topic,
+		.message.payload.data = msg->payload.ptr,
+		.message.payload.len = msg->payload.size,
+		.message.topic.qos = msg->qos,
+		.message_id = msg->message_id,
+		.dup_flag = msg->dup_flag,
+		.retain_flag = msg->retain_flag,
 	};
 
-	if (!connection_state_verify(STATE_CONNECTED)) {
+	if (!iot_hub_state_verify(STATE_CONNECTED)) {
 		LOG_WRN("Azure IoT Hub is not connected");
 		return -ENOTCONN;
 	}
 
-	len = snprintk(topic, sizeof(topic), TOPIC_DIRECT_METHOD_RES,
-			result->status, result->rid);
-	if ((len < 0) || (len > sizeof(topic))) {
-		LOG_ERR("Failed to create method result topic");
-		return -ENOMEM;
-	}
-
-	param.message.topic.topic.size = len;
+	switch (msg->topic.type) {
+	case AZURE_IOT_HUB_TOPIC_EVENT: {
+		az_iot_message_properties az_properties;
+		size_t property_count = (size_t)msg->topic.property_count;
+		// TODO: Properties are already allocated by the caller, possible to use that buffer
```

Comment:
All those TODOs, if they stay, they should use C-style comments.

Reply:
Removed the TODOs I could find.